### PR TITLE
[WIP] [Exploratory] Profiling pipeline part1

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

File: aten/src/ATen/core/jit_type.h [hunk 1]
```diff
@@ -391,6 +392,239 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
   int64_t dim_;
 };
 
+
+template <typename T>
+inline c10::optional<T> merge_primitive (const c10::optional<T>& a, const c10::optional<T>& b)
+{
```
Comment (bwasti) [inline aten/src/ATen/core/jit_type.h]: nit: are you using the `.clang-format` at the root of pytorch?

Comment (Krovatkin) [reply aten/src/ATen/core/jit_type.h]: I usually use `git-clang-format`, so it formats only my changes, but for this PR, I haven't run it yet, since I anticipate that Zach and you will want me to move stuff around.

File: torch/csrc/jit/interpreter.cpp [hunk 2]
```diff
@@ -412,6 +412,21 @@ struct CodeImpl {
     for (auto node : block->nodes()) {
       const auto& source_location = node->getSourceLocation();
       switch (node->kind()) {
+        case prim::profile:
```
Comment (bwasti) [inline torch/csrc/jit/interpreter.cpp]: do we need profiling options to be in the IR itself? can we get away with a simple map/set of Values with profiling options?

File: torch/csrc/jit/ir.cpp [hunk 3]
```diff
@@ -584,7 +584,13 @@ void Block::reIndexTopology() {
...
 }
 
-void Block::cloneFrom(Block* src, std::function<Value*(Value*)> value_map) {
+
+void Block::cloneFrom(Block* src, std::function<Value*(Value*)> value_map)
+{
+  cloneWithProfile(src, value_map, false);
+}
+
+void Block::cloneWithProfile(Block* src, std::function<Value*(Value*)> value_map, bool profile) {
```
Comment (bwasti) [inline torch/csrc/jit/ir.cpp]: should this also copy the profiling count?

File: torch/csrc/jit/ir.h [hunk 4]
```diff
@@ -1294,5 +1315,25 @@ TORCH_API std::vector<Value*> inlineCallTo(
...
+struct ProfilingRecord {
+
+
+  ProfilingRecord(const std::shared_ptr<Graph>& g): profiled_graph_(g) {}
+  ProfilingRecord(const ProfilingRecord& other);
+
+  std::unique_ptr<ProfilingRecord> copy () const;
+  std::unordered_map<size_t, ProfiledTensorTypePtr> profiled_types_;
+  std::shared_ptr<Graph> profiled_graph_;
+  void record (const IValue& ival, size_t unique);
```
Comment (bwasti) [inline torch/csrc/jit/ir.h]: I think you'll want to record Value -> IValue?

File: torch/csrc/jit/interpreter.cpp [hunk 5]
```diff
@@ -412,6 +412,21 @@ struct CodeImpl {
...
+        {
+          Operation op = [this, node](Stack& stack) -> int
+          {
+            IValue t;
+            pop(stack, t);
+
+            if (t.isTensor())
+            {
+                this->profiling_record_->record(t, static_cast<size_t>(node->i(attr::value)));
```
Comment (bwasti) [inline torch/csrc/jit/interpreter.cpp]: casting to size_t for a unique ID seems funky when we have unique addressable `Value` nodes

File: torch/csrc/jit/ir.h [hunk 6]
```diff
@@ -934,11 +940,17 @@ struct Block {
       owning_node_; // either the node that has this block or nullptr for root
 };
 
+struct SourceGraph {
```
Comment (bwasti) [inline torch/csrc/jit/ir.h]: what is this for?

**Review by bwasti — CHANGES_REQUESTED** [review]
The structures (ProfiledTensorType, ProfileRecord) make sense given what we discussed.  However, I am a bit skeptical that the profiling changes need to invasively change the graph implementation.  Ideally this can be done without any modifications to the IR and instead be some kind of object (that is similar to ProfileRecord) passed into the interpreter.

I've left various comments, but some might be made irrelevant if the design changes to avoid changing Graph.

File: aten/src/ATen/core/jit_type.h [hunk 7]
```diff
@@ -391,6 +392,239 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
...
+    return ProfiledTensorTypePtr(new ProfiledTensorType(scalar_type, device, sizes, strides, requires_grad)); // NOLINT(modernize-make-shared)
+  }
+
+  static ProfiledTensorTypePtr create(c10::optional<at::ScalarType> scalar_type, c10::optional<Device> device,const VaryingShape& sizes, const VaryingStrides& strides, c10::optional<bool> requires_grad)
+  {
+      return ProfiledTensorTypePtr(new ProfiledTensorType(scalar_type, device, sizes, strides, requires_grad)); // NOLINT(modernize-make-shared)
+  }
+
+
+  const VaryingShape& sizes() const { return sizes_; }
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply aten/src/ATen/core/jit_type.h]: We probably could. Unfortunately, we also need to worry about `strides_` which are similar to `sizes_` (they actually will probably need even more customization to be useful, which I think, we could add via subclassing `VaryingShape`) and I don't want to end up with `ProfiledTensorType`'s API that looks something like this: `getSizeDimensionAt`, `getStrideAt`. Also, it will be inconsistent with `CompleteTensorType` which `sizes()` returns a collection. How strongly do you feel about this?

File: aten/src/ATen/core/jit_type.h [hunk 8]
```diff
@@ -391,6 +392,239 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
...
+           TensorType::isSubtypeOf(rhs);
+  }
+  bool isSubclass(const TypeKind kind) const override {
+    return kind == TypeKind::TensorType ||
+           kind == TypeKind::ProfiledTensorType;
+  }
+  std::string str() const override {
+    // str is used for user-facing error messages, where we
+    // don't want to reveal underlying size information.
+    return "Tensor";
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply aten/src/ATen/core/jit_type.h]: We could definitely make this more descriptive, although we aren't planning to expose this to users in the near future.

File: torch/csrc/jit/graph_executor.cpp [hunk 9]
```diff
@@ -56,7 +56,29 @@ struct ExecutionPlan {
...
+      // another strategy is to keep a record per run
+      pr =  profiling_records_.size() > 0 ? std::move(profiling_records_[0]) : make_unique<ProfilingRecord>(graph);
+      is.setProfilingRecord(std::move(pr));
+    }
+    is.run(stack);
+    if (profile)
+    {
+      graph->setProfilingCount(graph->getProfilingCount() - 1);
+      pr = is.getProfilingRecord();
+      pr->dump();
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/graph_executor.cpp]: ah! thank you!

File: torch/csrc/jit/graph_executor.cpp [hunk 10]
```diff
@@ -56,7 +56,29 @@ struct ExecutionPlan {
...
+      profiling_records_.emplace_back(pr.release());
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/graph_executor.cpp]: `emplace_back` seems to consume constructor's **arguments**, directly (a raw pointer in our case), it's kinda similar to `make_shared`, `make_unique`. I believe, `profiling_records_.push_back(std::move(pr))` doesn't work either, but I'll double check.

File: torch/csrc/jit/interpreter.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/interpreter.cpp]: We can, but it ends up being a way more complicated and way less flexible. I have another branch where I explored that option. Interpreter runs a *different* graph from the one we want to collect profiling info (stored in an `ExecutionPlan`), so we will need to keep a map from the values in a new copied graph back to the values in the original graph whereas with profile instructions we have a flexibility to use the old value ids or (any kind of integral identifier for that matter) directly. They also have an added benefit of interfering with optimizations, so we know that values won't disappear on us, although this is less of an issue as of now since we are profiling the optimized version of a graph.

File: torch/csrc/jit/interpreter.cpp (hunk 5 above)
Comment (Krovatkin) [reply torch/csrc/jit/interpreter.cpp]: They have very nice properties, though. They are consecutive. They should remain valid after graph copies if the ids in an original graph were consecutive. They should be serializable and deserializable. They are more versatile in a sense we could use them in IR as well, and in different places, etc

File: torch/csrc/jit/interpreter.cpp [hunk 11]
```diff
@@ -857,7 +891,7 @@ std::ostream& operator<<(std::ostream& out, const Code& code) {
   return out;
 }
 
-Code::Code(const std::shared_ptr<Graph>& graph) : pImpl(new CodeImpl(graph)) {}
+Code::Code(const std::shared_ptr<Graph>& graph_) : pImpl(new CodeImpl(graph_)) {}
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/interpreter.cpp]: I know... weird... I blame Atom 😄 ... and myself for not fixing it in one of the clean-up passes.

File: torch/csrc/jit/interpreter.cpp [hunk 12]
```diff
@@ -846,6 +865,21 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
...
+
+  //this tells Interpreter to start profiling
+  void setProfilingRecord(std::unique_ptr<ProfilingRecord> profiling_record)
+  {
+    profiling_record_ = std::move(profiling_record);
+  }
+
+  std::unique_ptr<ProfilingRecord> getProfilingRecord()
+  {
+    return std::move(profiling_record_);
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/interpreter.cpp]: will fix

File: torch/csrc/jit/ir.cpp (hunk 3 above)
Comment (Krovatkin) [reply torch/csrc/jit/ir.cpp]: This does: https://github.com/pytorch/pytorch/pull/18772/files#diff-07e06aa4afc5f889bb134d6d6415ad2aR658
Frankly, I'm not very excited about copying and carrying around a profiling count, but it might be useful to know that this graph is being profiled.

File: torch/csrc/jit/ir.h (hunk 6 above)
Comment (Krovatkin) [reply torch/csrc/jit/ir.h]: arrgh, will remove! `SourceGraph` was used in another version w/o profiling instructions.

File: torch/csrc/jit/ir.h (hunk 4 above)
Comment (Krovatkin) [reply torch/csrc/jit/ir.h]: `unique -> value` is more flexible (described in the other comment). A consumer of profiling information will still pass `Value*` from which we extract its unique id and locate the corresponding profiling info.

**Comment by apaszke** [comment]
Is there any writeup of those discussions? I'd be very interested to see what's the plan for this.

**Comment by Krovatkin** [comment]
@apaszke not yet, but there's probably been enough discussions to start formalizing the goals and some implementation details/strategies. I'll try to put one together in the next few days.

**Review by zdevito — COMMENTED** [review]
Not a full review. Some thoughts:

1. The ProfilingRecord should be its own component separate from ir.h, and interpreter.h
2. Execution engine should be managing ProfileRecord objects for its optimized graphs. It should call a pass to add profile information.
3. Be careful about fork/wait primitives. Just because interpreter has run does not mean it has finished running the code.
4. There shouldn't need to be changes to interpreter.h or ir.h, I'd expect most things to be either new files or changes that are in execution_engine.cpp
5. We should be clear how we distinguish time profiling for operators, shape profiling in the execution engine, and quantization profiling. They are three separate systems that have some similar properties. Over time they may end up sharing code. At the very least `prim::profile` will need a more specific name.

File: torch/csrc/jit/pybind_utils.h [hunk 13]
```diff
@@ -235,6 +235,11 @@ inline IValue toIValue(
       }
       return userObj;
     }
+    case TypeKind::ProfiledTensorType:
```
Comment (zdevito) [inline torch/csrc/jit/pybind_utils.h]: nit: same is true of DimensionedTensorType and CompleteTensorType but they are handled differently here. This needs to be consistent, probably by making ProfiledTensorType behave the same as DimensionedTensorType here.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 14]
```diff
@@ -111,6 +112,26 @@ static at::Tensor to_dispatch(
 
 RegisterOperators reg(
     {Operator(
+         "prim::profile(Tensor t) -> ()",
+         [](const Node* node) {
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: Capture precisely what is need, not the entire node. Technically we do not require the Graph to stay alive after creating code. Keep in mind this is the interpreter's internal graph, it has not responsibility to keep it alive. In this case both the ProfilingRecord and the Value* can be determined here and then captured by the lambda.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 15]
```diff
@@ -111,6 +112,26 @@ static at::Tensor to_dispatch(
...
+         "prim::finish_profile() -> ()",
+         [](const Node* node) {
+           return [node](Stack& stack) {
+             auto p_pr =
+                 reinterpret_cast<ProfilingRecord*>(node->i(attr::storage));
+             p_pr->finish(node, stack);
+             return 0;
+           };
+         }),
+     Operator(
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: I still think this might be cleaner if the only registered op was:

```
	     Operator(
         "prim::observe(...) -> ()",
         [](const Node* node) {
           size_t N = node->inputs().size();
           std::function<void(at::ArrayRef<IValue>)>* callback = reinterpret_cast<...>(node->i(attr::callback));
           return [node](Stack& stack) {
                (*callback)(last(stack, N));
             return 0;
           };
         }),
```
It would make it easier to change what the profiler does with observations in the future since the linking between the observed value and where to record it stays inside record profile. It would be  harder to that with the current implementation because it is unsafe to capture `node` as described above. I also am not sure that all observers really can fit as subclasses to ProfilingRecord. For instance, not all profilers would be recording information into a Graph, but ProfilingRecord always has a graph.

File: torch/csrc/jit/profiling_record.cpp [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    for (auto b : n->blocks()) {
+      instrumentBlock(b);
+    }
+  }
+}
+
+std::shared_ptr<Graph> ProfilingRecord::instrumentGraph(
+    const std::shared_ptr<Graph>& graph) {
+  auto new_g = graph->copy();
+  profiled_graph_ = new_g;
```
Comment (zdevito) [inline torch/csrc/jit/profiling_record.cpp]: weird -- I would not expect this method as it is named to delete other profiled graphs. Repeated calls to this method are pretty non-sensical and unsafe! What happens if we continue to record into the old profiled_graph_ after we switched it over.

**Review by zdevito — COMMENTED** [review]
This is pretty cool! I have a bunch of inline comments, but they should be easy to address.

File: aten/src/ATen/core/jit_type.h [hunk 17]
```diff
@@ -391,6 +392,231 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
...
+  c10::optional<bool> requiresGrad() const { return requires_grad_; }
+
+  ProfiledTensorTypePtr toScalarType(at::ScalarType type){
+    auto t = ProfiledTensorType::create(*this);
+    t->scalar_type_ = type;
+    return t;
+  }
+
+  bool operator==(const Type& rhs) const override {
+    if(rhs.kind() != kind())
```
Comment (ZolotukhinM) [inline aten/src/ATen/core/jit_type.h]: Styling nit: single-statement ifs still need {}-braces.

File: torch/csrc/jit/graph_executor.cpp [hunk 18]
```diff
@@ -379,7 +380,7 @@ struct GraphExecutorImpl {
 
     auto& execution_plan =
         optimize ? getOrCompile(stack) : getOrCompileFallback();
-    return execution_plan.run(stack);
+    execution_plan.run(stack);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/graph_executor.cpp]: The change seems unrelated.

File: torch/csrc/jit/graph_executor.cpp [hunk 19]
```diff
@@ -441,11 +442,23 @@ struct GraphExecutorImpl {
...
+        it = plan_cache.emplace(spec, std::move(plan)).first;
+      }
+
+      auto& plan = it->second;
+      if (profiling_record_.profiling_count_ == 0) {
+        //re-compile with profiling info
+        auto new_plan = compileSpec(spec);
+        plan_cache[spec] = std::move(plan);
+        profiling_record_.profiling_count_ = -1;
+        return plan_cache[spec];
```
Comment (ZolotukhinM) [inline torch/csrc/jit/graph_executor.cpp]: Nit: can return `plan_cache[spec]` in both `if` and `else`, or even better: remove the `else`, return after the `if`.

File: torch/csrc/jit/profiling_record.cpp [hunk 20]
```diff
@@ -0,0 +1,118 @@
+#include <jit/profiling_record.h>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/profiling_record.cpp]: Nit: usually jit headers are included with a more complete path: `torch/csrc/jit/...`

File: torch/csrc/jit/profiling_record.cpp [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+Node* ProfilingRecord::createProfileNode(Graph* graph, Value* o) {
+  Node* pop = graph->create(prim::profile, 0);
+  pop->i_(attr::data, reinterpret_cast<int64_t>(o));
+  addProfilingRecordAttribute(pop);
+  return pop;
+}
+
+void ProfilingRecord::instrumentBlock(Block* block) {
+  auto graph = block->owningGraph();
+  for (auto it = block->nodes().rend(); it != block->nodes().rbegin(); --it) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/profiling_record.cpp]: Do we go backwards to avoid instrumenting `::profile` nodes? Please add a comment.

File: aten/src/ATen/core/jit_type.h (hunk 1 above)
Comment (ZolotukhinM) [reply aten/src/ATen/core/jit_type.h]: Please run it though :)

**Review by ZolotukhinM — COMMENTED** [review]
Some more minor remarks from me :) I'll need another pass through the code to better understand it (if you could address the current remarks, it would make it easier!).

File: aten/src/ATen/core/jit_type.h [hunk 22]
```diff
@@ -391,6 +392,231 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
...
+    ProfiledTensorType(c10::optional<at::ScalarType> scalar_type, c10::optional<Device> device,const VaryingShape& sizes, const VaryingStrides& strides, c10::optional<bool> requires_grad)
+    : TensorType()
+    , scalar_type_(scalar_type)
+    , device_(device)
+    , sizes_(sizes)
+    , strides_(strides)
+    , requires_grad_(requires_grad)
+    {}
+
+    static std::vector<int64_t> contiguousStridesOf(at::IntArrayRef sizes) {
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply aten/src/ATen/core/jit_type.h]: removed

File: aten/src/ATen/core/type.cpp [hunk 23]
```diff
@@ -23,6 +23,17 @@ std::ostream& operator<<(std::ostream & out, const Type & t) {
       }
     }
     out << ")";
+  } else if (auto value = t.cast<ProfiledTensorType>()) {
+    out << "Tensor(dtype = ";
+    if  (value->scalarType().has_value())
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply aten/src/ATen/core/type.cpp]: I want it to print out "dynamic."

File: torch/csrc/jit/profiling_record.cpp (hunk 21 above)
Comment (Krovatkin) [reply torch/csrc/jit/profiling_record.cpp]: added

File: torch/csrc/jit/register_prim_ops.cpp (hunk 14 above)
Comment (Krovatkin) [reply torch/csrc/jit/register_prim_ops.cpp]: fixed

File: torch/csrc/jit/register_prim_ops.cpp (hunk 15 above)
Comment (Krovatkin) [reply torch/csrc/jit/register_prim_ops.cpp]: fixed

File: aten/src/ATen/core/jit_type.h [hunk 24]
```diff
@@ -391,6 +392,231 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
...
+{
+  if (a.has_value() && b.has_value() &&
+      a.value() == b.value()) {
+    return a;
+  }
+  return c10::optional<T>{};
+}
+
+
+struct CAFFE2_API VaryingShape {
```
[the remark this answers is not in the document]
Comment (zdevito) [reply aten/src/ATen/core/jit_type.h]: nit: Sentence one says "varying dimension/ranks" which I can infer from the name of the class. Sentence two tells me about the API which I can see from the type signatures. But is missing is something like: This is a list of numbers, where each number is optional. If it is not present it means that dimension has been different across runs. If the entire list is not present, that means the _rank_ varies across runs.

File: aten/src/ATen/core/jit_type.h [hunk 25]
```diff
@@ -391,6 +392,231 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
...
+    auto rt = rhs.expect<ProfiledTensorType>();
+    return scalar_type_ == rt->scalarType() &&
+           sizes() == rt->sizes() &&
+           strides() == rt->strides() &&
+           device() == rt->device();
+  }
+  bool isSubtypeOf(const TypePtr rhs) const override {
+    if (rhs->kind() == TypeKind::ProfiledTensorType)
+      return *expect<ProfiledTensorType>() == *rhs;
+    return rhs->kind() == TypeKind::TensorType ||
```
[the remark this answers is not in the document]
Comment (zdevito) [reply aten/src/ATen/core/jit_type.h]: Can you comment here?

File: aten/src/ATen/core/type.cpp [hunk 26]
```diff
@@ -486,4 +497,49 @@ void ClassType::clearRegistry() {
   getRegistry().clear();
 }
 
+std::string ProfiledTensorType::str() const {
```
Comment (zdevito) [inline aten/src/ATen/core/type.cpp]: Did we need an override for this? Doesn't the fallback to TensorType handle it?

File: torch/csrc/jit/ir.h [hunk 27]
```diff
@@ -968,6 +968,10 @@ struct Graph {
 
   Graph() : Graph(c10::make_intrusive<Scope>()) {}
 
+  const std::unordered_set<const Value*>& values() {
```
Comment (zdevito) [inline torch/csrc/jit/ir.h]: Code is dead as far as I can tell. This must be deleted. If left in, it will lead to a bunch of people misusing it in incredibly awful ways. They will forget that Blocks may exist. They will forget the inputs to the graph will exist. They will forget the order of this thing is undetermined and create a bunch of non-deterministic code.

File: torch/csrc/jit/profiling_record.cpp [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+
+  // iterating backwards allows us to easily insert profile nodes
+  // without affecting an iterator
+  for (auto it = block->nodes().rend(); it != block->nodes().rbegin(); --it) {
+    auto n = *it;
+    for (auto o : n->outputs()) {
+      if (!o->type()->isSubclass(TypeKind::TensorType)) {
+        continue;
+      }
+      auto pn = graph->create(prim::profile, 0);
```
Comment (zdevito) [inline torch/csrc/jit/profiling_record.cpp]: nit: I feel like the way this is organized into functions is weird. I would have expected:

```
Node* ProfiledTensorType::createCallback(std::function<void(Stack&)> callback, at::ArrayRef<Value> inputs);
```
Rather than addCallback. That way the details of a prim::profile node are contained in one function.

File: torch/csrc/jit/profiling_record.cpp [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+  auto pr = std::unique_ptr<ProfilingRecord>(new ProfilingRecord(new_g));
+  auto raw_pr = pr.get();
+
+  pr->instrumentBlock(new_g->block());
+  std::function<void(Stack&)> counter = [raw_pr](Stack&) {
+    std::lock_guard<std::mutex> lock(raw_pr->mutex_);
+    raw_pr->profiling_count_--;
+  };
+
+  auto pop = new_g->create(prim::profile, 0);
```
Comment (zdevito) [inline torch/csrc/jit/profiling_record.cpp]: note how the details (attribute name, inputs, etc.) of prim::profile get repeated here.

**Review by zdevito — COMMENTED** [review]
Looks good! There is one line that I need to make sure is deleted, I will approve once it is.

**Review by zdevito — APPROVED** [review]

File: test/cpp/jit/test_misc.h [hunk 30]
```diff
@@ -748,6 +749,56 @@ graph(%a):
   AT_ASSERT(testPassValue);
 }
 
+static void checkShape(Node* n, std::vector<int64_t> expected) {
+  auto tp = n->output()->type();
+  auto ptp = tp->expect<ProfiledTensorType>();
+  ASSERT_EQ(ptp->sizes().concrete_sizes().value(), expected);
+}
+
+void testProfiler() {
```
Comment (ZolotukhinM) [inline test/cpp/jit/test_misc.h]: Do we really need an entire LSTM to test a profiler? Could we get a smaller test? It's a bit hard to tell what the profiler does from this test.

File: aten/src/ATen/core/jit_type.h [hunk 31]
```diff
@@ -736,7 +919,10 @@ struct CAFFE2_API TupleType : public Type {
 
   bool compare(const Type& rhs, std::function<bool(const TypePtr, const TypePtr)> fn) const {
     if(rhs.kind() != kind())
+    {
```
Comment (ZolotukhinM) [inline aten/src/ATen/core/jit_type.h]: While I appreciate that you fix this, I'd prefer this to be committed separately (along with a similar fix `s/bool numel/size_t numel/` above and other cleanups if any).

File: aten/src/ATen/core/jit_type.h [hunk 32]
```diff
@@ -391,6 +392,231 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
...
+      }
+      frozen.push_back(d.value());
+    }
+
+    return frozen;
+  }
+
+private:
+  bool is_varying_size_;
+  std::vector<c10::optional<int64_t>> dims_;
```
[the remark this answers is not in the document]
Comment (ZolotukhinM) [reply aten/src/ATen/core/jit_type.h]: My (slight) preference goes to `optional<vector<optional<int>>>`. It might not look beautiful, but it uses `optional` in a use case it was designed for and does not reimplement it in a custom way. IOW, in the current code a similar problem is solved in two different ways in the same place, which I don't really like that much.

File: torch/csrc/jit/profiling_record.h [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+
+#include <list>
+#include <vector>
+
+namespace torch {
+namespace jit {
+
+using ::c10::ProfiledTensorTypePtr;
+
+struct ProfilingRecord {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/profiling_record.h]: Please add a top-level comment.

File: torch/csrc/jit/profiling_record.h [hunk 34]
```diff
@@ -0,0 +1,44 @@
...
+  // are captured in callbacks_
+  ProfilingRecord(const ProfilingRecord&) = delete;
+  ProfilingRecord(ProfilingRecord&&) noexcept = delete;
+  static ProfiledTensorTypePtr toProfiledTensorTypePtr(const IValue& ival);
+  static std::unique_ptr<ProfilingRecord> instrumentGraph(
+      const std::shared_ptr<Graph>& graph);
+
+  std::shared_ptr<Graph> profiled_graph_;
+  std::mutex mutex_;
+  size_t profiling_count_;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/profiling_record.h]: Shouldn't this be private?

File: torch/csrc/jit/profiling_record.h (hunk 33 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/profiling_record.h]: Also, I think the struct needs `TORCH_API` on the declaration.

**Review by ZolotukhinM — COMMENTED** [review]
Please find some remarks from me inline (mostly styling).
